### PR TITLE
Secrets

### PR DESCRIPTION
**Pull request by djmb**
A big overhaul of how secrets are handled by Kamal.

1. We'll read them in from `.kamal/secrets[.<destination>]` instead of `.env`. This prevents clashes with other tools that use .env.
2. No more `kamal envify` or `kamal env push` - secrets are either generated externally to Kamal, or can be interpolated at runtime by dotenv's [command and variable substitution](https://github.com/bkeepers/dotenv?tab=readme-ov-file#command-substitution)
3. We don't load the secrets into Kamal's environment, so you can't refer to them in deploy.yml ERB
4. Secrets are loaded lazily, so commands that don't need them can be run without the secrets file.
5. There's a new `kamal secrets` subcommand that provides wrappers for common password management tools (right now 1Password, LastPass and Bitwarden)
6. There's no separate env file for secrets anymore, they are passed on the command line with all the other env variables. We expect them to be up to date at all times.

~~Passing the secrets on the command line means they'll end up in the process list. We want to avoid so this PR will be followed up with one that generates and uploads a single env file with the full env instead.~~

Update: secrets are now uploaded to an env file on the server, to avoid them ending up in process lists.

**kamal secrets**

```
kamal secrets --help
Commands:
  kamal secrets extract                                                     # Extract a single secret from the results of a fetch call
  kamal secrets fetch [SECRETS...] --account=ACCOUNT -a, --adapter=ADAPTER  # Fetch secrets from a vault
  kamal secrets help [COMMAND]                                              # Describe subcommands or one specific subcommand

Options:
...
```

`kamal secrets fetch` returns a JSON blob of secrets for the chosen adapter, e.g.:

```
kamal secrets fetch --adapter 1password --account myaccount --from op://MyVault/MyItem production/RAILS_MASTER_KEY production/DB_PASSWORD common/ANNOUNCEMENT_TOKEN
```

Then you can pass that JSON blob to `kamal secrets extract` to load a single secret.

```
# These all work
kamal secrets extract DB_PASSWORD <JSON>
kamal secrets extract production/DB_PASSWORD <JSON>
kamal secrets extract MyItem/production/DB_PASSWORD <JSON>
kamal secrets extract MyVault/MyItem/production/DB_PASSWORD <JSON>
```

You can combine these to create a secrets file like this:
```
# .kamal/secrets
SECRETS=$(kamal secrets fetch --adapter 1password --account myaccount --from op://MyVault/MyItem production/RAILS_MASTER_KEY production/DB_PASSWORD common/ANNOUNCEMENT_TOKEN)

RAILS_MASTER_KEY=$(kamal secrets extract RAILS_MASTER_KEY ${SECRETS})
DB_PASSWORD=$(kamal secrets extract DB_PASSWORD ${SECRETS})
ANNOUNCEMENT_TOKEN=$(kamal secrets extract ANNOUNCEMENT_TOKEN ${SECRETS})
```

The single `kamal secrets fetch` call allows the adapter to fetch the secrets as efficiently as possible, so for example in 1Password, if multiple secrets are stored in a single item we only need to load it once.

### CODE REVIEW

File: lib/kamal/secrets/adapters/bitwarden.rb [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+        end
+      end
+    end
+
+    def signedin?(account)
+      run_command("status")["status"] != "unauthenticated"
+    end
+
+    def run_command(command, session: nil, raw: false)
+      full_command = [ *("BW_SESSION=#{session}" if session), "bw", command ].join(" ")
```
[the remark this answers is not in the document]
Comment (djmb) [reply lib/kamal/secrets/adapters/bitwarden.rb:60]: You should be providing the input yourself as a user of kamal, but I'll fix these up anyway.

**Comment by nickhammond** [comment]
Does this PR also mean that every deploy will need to have access to the secret store at run/deploy time? e.g. in your example that would be 4 CLI calls to the `op` CLI?

File: lib/kamal/cli/templates/secrets [hunk 2]
```diff
@@ -0,0 +1,6 @@
+# SECRETS=$(kamal secrets --adapter 1password --from Vault/Item Section1/KAMAL_REGISTRY_PASSWORD Section2/RAILS_MASTER_KEY)
+# KAMAL_REGISTRY_PASSWORD=$(kamal secrets extract KAMAL_REGISTRY_PASSWORD ${SECRETS})
+# KAMAL_REGISTRY_PASSWORD=$(kamal secrets extract RAILS_MASTER_KEY ${SECRETS})
```
Comment (Sija) [inline lib/kamal/cli/templates/secrets:3]:
```suggestion
# RAILS_MASTER_KEY=$(kamal secrets extract RAILS_MASTER_KEY ${SECRETS})
```

**Review by Sija — CHANGES_REQUESTED** [review]

**Comment by djmb** [comment]
@nickhammond - yes you will need access a runtime if you are interpolating the secrets. You could generate .kamal/secrets separately (like `kamal envify` currently does), but that's something you'd need to manage externally to kamal.

In the example below there's only a single call to op:
```
SECRETS=$(kamal secrets fetch --adapter 1password --account myaccount --from op://MyVault/MyItem production/RAILS_MASTER_KEY production/DB_PASSWORD common/ANNOUNCEMENT_TOKEN)

RAILS_MASTER_KEY=$(kamal secrets extract RAILS_MASTER_KEY ${SECRETS})
DB_PASSWORD=$(kamal secrets extract DB_PASSWORD ${SECRETS})
ANNOUNCEMENT_TOKEN=$(kamal secrets extract ANNOUNCEMENT_TOKEN ${SECRETS})
```

The `kamal secrets fetch` call will check you are logged in and prompt you for your 1password password if not, then it will make a single call to read op://MyVault/MyItem. Something like:

```
op item get --vault MyVault MyItem --format json --fields label=production.RAILS_MASTER_KEY,label=production.DB_PASSWORD,label=common.ANNOUNCEMENT_TOKEN
```

It returns a JSON string which is a hash of secret key to value and this gets saved in the SECRETS ENV variable. `kamal secrets extract` just extracts the value from the variable without calling `op`.

For the 1Password adapter, you can get multiple items from single `kamal secrets fetch` call, but they'll be read with multiple 
`op item get` calls.

E.g. :
```
kamal secrets fetch --adapter 1password --account myaccount --from op://MyVault MyItem1/production/RAILS_MASTER_KEY MyItem1/production/DB_PASSWORD MyItem2/common/ANNOUNCEMENT_TOKEN
```

**Comment by mhenrixon** [comment]
Hej @djmb would it be possible to keep the inheritance so that I can avoid duplicating environment variables for each environment? I have an app that deploys to many different environments. It was phenomenal to prevent duplicating the kamal registry username/password. 

It isn't a huge deal but it was convenient. I tried putting it in `.kamal/secrets`, but that didn't get picked up.

**Comment by djmb** [comment]
@mhenrixon - yes that should be possible 👍

**Pull request merged**